File: acmicpc/4949/4949.cc
#include <cstdio>
#include <cstdlib>
#include <cstring>

void solveline(char *s) {
    char stack[102];
    int idx = 0;
    while (*s != '\0') {
        char c = *s;
        if (c == '(' || c== '[') {
            stack[idx++] = c;
        } else if (c==')'||c==']'){
            if (idx > 0 && 
                ((stack[idx-1] == '(' && c== ')') ||
                (stack[idx-1] == '[' && c== ']'))) {
                idx--;
            }  
            else {
                printf("no\n");
                return;
            }
        }
        s++;
    }
    if (idx != 0) {
        printf("no\n");
    } else {
        printf("yes\n");
    }
}

int main() {
    char *s = NULL;
    size_t len = 0;
    for (;;) {
        getline(&s, &len, stdin);
        if (strcmp(s, ".") == 0 || strcmp(s, ".\n") == 0) {
            break;
        }
        
        solveline(s);
    }
    free(s);
}

File: acmicpc/4949/4949.py

def solve_line(line):
    stack = []

    for s in line:
        if s == '(' or s== '[':
            stack.append(s)
        elif s == ')' or s == ']':
            empty = len(stack) != 0
            if empty and s == ')' and stack[-1] == '(':
                stack.pop()
            elif empty and s == ']' and stack[-1] == '[':
                stack.pop()
            else:
                print("no")
                return
    if len(stack) != 0:
        print("no")
    else:
        print("yes")
        


if __name__ == "__main__":
    while True:
        line = input()
        if line == ".":
            break
        else:
            solve_line(line)